syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+	int argvID = 0;
+
+	char *imaps, *dmaps;
+	unsigned int flags = 0;
+
+	const char *appArgv;
+	char name[SIZE_CMD_ARG_LINE];
+
+	if (argc < 5 || argc > 6) {
+		fprintf(stderr, "\n%s: Wrong argument count", argv[0]);
+		return -1;
+	}
+
+	/* ARG_0: command name */
+
+	/* ARG_1: alias to device - it will be checked in phfs_open */
+
+	/* ARG_2: optional flags */
+	argvID = 2;
+	if (argv[argvID][0] == '-') {
+		if ((argv[argvID][1] | 0x20) == 'x' && argv[argvID][2] == '\0') {
+			flags |= flagSyspageExec;
+			argvID++;
+		}
+		else {
+			fprintf(stderr, "\n%s: Wrong arguments", argv[0]);
+			return -1;
+		}
+	}
+
+	if (argvID != (argc - 3)) {
+		fprintf(stderr, "\n%s: Invalid arg, 'dmap' is not declared", argv[0]);
+		return -1;
+	}
+
+	/* ARG_3: name + argv */
+	appArgv = argv[argvID];
+	for (pos = 0; appArgv[pos]; pos++) {
+		if (appArgv[pos] == ';')
+			break;
+	}
+	memcpy(name, appArgv, pos);
+	name[pos] = '\0';
+
+	/* ARG_4: maps for instructions */
+	imaps = argv[++argvID];
+
+	/* ARG_5: maps for data */
+	dmaps = argv[++argvID];
+
+	return sysgen_appAdd(name, imaps, dmaps, appArgv, flags);
+}
+
+
+int sysgen_cmdConsole(int argc, char *argv[])
+{
+	char *endptr;
+	unsigned int minor;
+
+	if (argc != 2) {
+		fprintf(stderr, "\n%s: Wrong argument count", argv[0]);
+		return -EINVAL;
+	}
+
+	strtoul(argv[1], &endptr, 0);
+	if (*endptr != '.') {
+		fprintf(stderr, "\nWrong major value: %s", argv[1]);
+		return -EINVAL;
+	}
+
+	minor = strtoul(++endptr, &endptr, 0);
+	if (*endptr != '\0') {
+		fprintf(stderr, "\nWrong minor value: %s", argv[1]);
+		return -EINVAL;
+	}
+
+	sysgen_common.syspage->console = minor;
+
+	return 0;
+}
+
+
+static int sysgen_parseArgLine(const char *lines, char *buf, size_t bufsz, char *argv[], size_t argvsz)
+{
+	char *const end = buf + bufsz;
+	int argc = 0;
+
+	if (lines == NULL || *lines == '\0')
+		return -1; /* EOF */
+
+	while (*lines != '\0') {
+		if (isspace(*lines)) {
+			if (isblank(*lines++))
+				continue;
+			else
+				break;
+		}
+
+		/* Argument count and one NULL pointer */
+		if (argc + 1 >= argvsz) {
+			fprintf(stderr, "\nToo many arguments");
+			return -EINVAL;
+		}
+
+		argv[argc++] = buf;
+
+		while (isgraph(*lines) && buf < end)
+			*buf++ = *lines++;
+
+		if (buf >= end) {
+			fprintf(stderr, "\nCommand buffer too small");
+			return -ENOMEM;
+		}
+
+		*buf++ = '\0';
+	}
+
+	argv[argc] = NULL;
+
+	return argc;
+}
+
+
+static int sysgen_parseScript(const char *fname)
+{
+	FILE *file;
+	size_t len = 0;
+	char *line = NULL;
+	int sz, argc, i, ret;
+	char *argv[SIZE_CMD_ARGV];
+	char argline[SIZE_CMD_ARG_LINE];
+
+	file = fopen(fname, "r");
+	if (file == NULL) {
+		fprintf(stderr, "Cannot open file %s\n", fname);
+		return -1;
+	}
+
+	while ((sz = getline(&line, &len, file)) != -1) {
+		argc = sysgen_parseArgLine(line, argline, SIZE_CMD_ARG_LINE, argv, SIZE_CMD_ARGV);
+		/* skip empty lines */
+		if (argc == 0) {
+			continue;
+		}
+		/* either end of script or no script */
+		else if (argc == -1) {
```

Comment:
My suggestion - move EOF check here and leave negative argc solely for errors

Reply:
Done